plugins/dynamic-discovery-backend/src/service/router.ts

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2024 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { errorHandler } from '@backstage/backend-common';
+import { FeatureMetadata, LoggerService } from '@backstage/backend-plugin-api';
+import express from 'express';
+import Router from 'express-promise-router';
+import { DynamicDiscoveryService } from '../interfaces/DynamicDiscoveryService';
+import { AlreadyRegisteredError } from '../implementations/discovery/errors';
+
+export interface RouterOptions {
+  logger: LoggerService;
+  discovery: DynamicDiscoveryService;
+}
+
+export async function createRouter(
+  options: RouterOptions,
+): Promise<express.Router> {
+  const { logger, discovery } = options;
+
+  const router = Router();
+  router.use(express.json());
+
+  router.get('/health', (_, response) => {
+    logger.info('PONG!');
+    response.json({ status: 'ok' });
+  });
+
+  router.post('/register', (req, res) => {
+    logger.info(`registering plugins ${JSON.stringify(req.body)}`);
+    const { instanceLocation, feature, featureLocation } = req.body as {
+      instanceLocation: { internalUrl: string; externalUrl: string };
+      feature: FeatureMetadata;
+      featureLocation: { internalUrl: string; externalUrl: string };
+    };
+    try {
+      discovery.register(instanceLocation, feature, featureLocation);
+    } catch (err) {
+      if (err instanceof AlreadyRegisteredError) {
```

Comment:
Avoid `instanceof` whenever possible, because it breaks if packages are duplicated. Do a string comparison on the name here instead.

Reply:
fixed in [aa51900](https://github.com/backstage/backstage/pull/24539/commits/aa519003aafdf3eee4d38f6f756c8ccd14f742db)